File: src/scrapers/poit_scraper.py
"""
POIT Scraper - Hämtar kungörelser från Post- och Inrikes Tidningar
Använder undetected-chromedriver för att kringgå bot-detection
"""

import os
import time
import logging
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta
from dataclasses import dataclass, asdict

import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait, Select
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException

logger = logging.getLogger(__name__)


@dataclass
class Kungorelse:
    """Representerar en kungörelse från POIT"""
    kungorelse_id: str
    rubrik: str
    amnesomrade: str
    publiceringsdatum: str
    organisationsnummer: Optional[str] = None
    foretag: Optional[str] = None
    innehall: Optional[str] = None
    url: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


class POITScraper:
    """
    Scraper för Post- och Inrikes Tidningar (POIT)

    Ämnesområden som är relevanta för företagsbevakning:
    - Aktiebolag (nybildning, ändring, likvidation)
    - Konkurser och företagsrekonstruktion
    - Fusion och delning
    - Kallelse på okända borgenärer
    """

    BASE_URL = "https://poit.bolagsverket.se/poit-app"
    SEARCH_URL = f"{BASE_URL}/sok"

    # Relevanta ämnesområden för företagsbevakning
    RELEVANTA_AMNESOMRADEN = [
        "Aktiebolag",
        "Konkurs",
        "Företagsrekonstruktion",
        "Fusion",
        "Delning av aktiebolag",
        "Likvidation",
        "Kallelse på okända borgenärer",
        "Ändring av bolagsordning",
    ]

    def __init__(
        self,
        headless: bool = True,
        nopecha_extension_path: Optional[str] = None,
        chrome_path: Optional[str] = None,
        timeout: int = 30
    ):
        self.headless = headless
        self.nopecha_extension_path = nopecha_extension_path
        self.chrome_path = chrome_path
        self.timeout = timeout
        self.driver: Optional[uc.Chrome] = None

    def _create_driver(self) -> uc.Chrome:
        """Skapar en undetected Chrome-instans"""
        options = uc.ChromeOptions()

        if self.headless:
            options.add_argument('--headless=new')

        # Standard options för stabilitet
        options.add_argument('--no-sandbox')
        options.add_argument('--disable-dev-shm-usage')
        options.add_argument('--disable-gpu')
        options.add_argument('--window-size=1920,1080')
        options.add_argument('--lang=sv-SE')

        # Lägg till NopeCHA extension om tillgänglig
        if self.nopecha_extension_path and os.path.exists(self.nopecha_extension_path):
            options.add_extension(self.nopecha_extension_path)
            logger.info(f"Laddar NopeCHA extension från: {self.nopecha_extension_path}")

        driver_kwargs = {
            'options': options,
            'use_subprocess': True,
        }

        if self.chrome_path:
            driver_kwargs['browser_executable_path'] = self.chrome_path

        return uc.Chrome(**driver_kwargs)

    def start(self):
        """Startar browser-sessionen"""
        if self.driver is None:
            logger.info("Startar undetected Chrome...")
            self.driver = self._create_driver()
            logger.info("Chrome startad")

    def stop(self):
        """Stänger browser-sessionen"""
        if self.driver:
            logger.info("Stänger Chrome...")
            self.driver.quit()
            self.driver = None

    def __enter__(self):
        self.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.stop()

    def _accept_cookies(self, max_wait: int = 15):
        """
        Accepterar cookie-dialogen om den visas.
        Väntar flexibelt upp till max_wait sekunder på att dialogen dyker upp.
        """
        wait = WebDriverWait(self.driver, max_wait)

        # Cookie-knapp selectors att försöka
        cookie_selectors = [
            "//button[contains(text(), 'OK, fortsätt')]",
            "//button[contains(text(), 'OK')]",
            "//button[contains(text(), 'Acceptera')]",
            "//button[contains(text(), 'Godkänn')]",
        ]

        # Vänta och leta efter cookie-knappen i upp till max_wait sekunder
        for attempt in range(max_wait):
            for selector in cookie_selectors:
                try:
                    cookie_btn = self.driver.find_element(By.XPATH, selector)
                    if cookie_btn.is_displayed():
                        # Använd JavaScript-klick för att undvika intercept-problem
                        self.driver.execute_script("arguments[0].click();", cookie_btn)
                        logger.info("Accepterade cookies")
                        time.sleep(1)
                        return True
                except NoSuchElementException:
                    continue

            # Vänta 1 sekund innan nästa försök
            time.sleep(1)

            # Kolla om vi redan är förbi cookie-dialogen
            try:
                page_text = self.driver.find_element(By.TAG_NAME, "body").text
                if "OK, fortsätt" not in page_text and "Vi använder cookies" not in page_text:
                    logger.debug("Cookie-dialog finns inte längre")
                    return True
            except:
                pass

        logger.debug(f"Ingen cookie-dialog hittades efter {max_wait} sekunder")
        return False

    def _wait_for_page_load(self, max_wait: int = 30):
        """Väntar på att sidan laddas och eventuell CAPTCHA löses"""
        wait = WebDriverWait(self.driver, max_wait)

        # Acceptera cookies först
        time.sleep(2)
        self._accept_cookies()

        # Vänta på att CAPTCHA försvinner (om den finns)
        captcha_selectors = [
            "//div[contains(@class, 'captcha')]",
            "//iframe[contains(@src, 'captcha')]",
            "//*[contains(text(), 'robot')]",
            "//*[contains(text(), 'What code is in the image')]",
        ]

        # Ge NopeCHA tid att lösa CAPTCHA
        time.sleep(2)

        # Kolla om vi fortfarande är på CAPTCHA-sidan
        for _ in range(max_wait // 3):
            page_text = self.driver.find_element(By.TAG_NAME, "body").text.lower()
            if "captcha" not in page_text and "robot" not in page_text:
                logger.info("Ingen CAPTCHA detekterad eller redan löst")
                return True
            logger.info("Väntar på CAPTCHA-lösning...")
            time.sleep(3)

        logger.warning("Timeout vid väntan på CAPTCHA")
        return False

    def search_by_orgnr(self, orgnr: str) -> List[Kungorelse]:
        """
        Söker kungörelser för ett specifikt organisationsnummer

        Args:
            orgnr: Organisationsnummer (10 siffror)

        Returns:
            Lista med kungörelser
        """
        self.start()
        results = []

        try:
            logger.info(f"Söker kungörelser för orgnr: {orgnr}")
            self.driver.get(self.SEARCH_URL)

            if not self._wait_for_page_load():
                logger.error("Kunde inte ladda söksidan")
                return results

            # Vänta på sökformuläret
            wait = WebDriverWait(self.driver, self.timeout)

            # Försök hitta och fylla i organisationsnummer
            try:
                # POIT har olika fältnamn beroende på söktyp
                orgnr_field = wait.until(
                    EC.presence_of_element_located((By.CSS_SELECTOR,
                        "input[name*='orgnr'], input[name*='organisationsnummer'], input[id*='orgnr']"))
                )
                orgnr_field.clear()
                orgnr_field.send_keys(orgnr)

                # Klicka på sök
                search_btn = self.driver.find_element(By.CSS_SELECTOR,
                    "button[type='submit'], input[type='submit'], button:contains('Sök')")
                search_btn.click()

                # Vänta på resultat
                time.sleep(2)

                # Parsa resultat
                results = self._parse_search_results()

            except TimeoutException:
                logger.warning(f"Kunde inte hitta sökfält för orgnr: {orgnr}")
            except NoSuchElementException as e:
                logger.warning(f"Element saknas: {e}")

        except Exception as e:
            logger.error(f"Fel vid sökning: {e}")

        return results

    def search_by_date_range(
        self,
        from_date: datetime,
        to_date: datetime,
        amnesomrade: Optional[str] = None
    ) -> List[Kungorelse]:
        """
        Söker kungörelser inom ett datumintervall

        Args:
            from_date: Startdatum
            to_date: Slutdatum
            amnesomrade: Filtrera på ämnesområde (optional)

        Returns:
            Lista med kungörelser
        """
        self.start()
        results = []

        try:
            logger.info(f"Söker kungörelser {from_date.date()} - {to_date.date()}")
            self.driver.get(self.SEARCH_URL)

            if not self._wait_for_page_load():
                logger.error("Kunde inte ladda söksidan")
                return results

            wait = WebDriverWait(self.driver, self.timeout)

            # Försök fylla i datumfält
            try:
                # Formatera datum för POIT (YYYY-MM-DD)
                from_str = from_date.strftime("%Y-%m-%d")
                to_str = to_date.strftime("%Y-%m-%d")

                # Hitta datumfält
                from_field = wait.until(
                    EC.presence_of_element_located((By.CSS_SELECTOR,
                        "input[name*='from'], input[name*='datum_fran'], input[id*='fromDate']"))
                )
                from_field.clear()
                from_field.send_keys(from_str)

                to_field = self.driver.find_element(By.CSS_SELECTOR,
                    "input[name*='to'], input[name*='datum_till'], input[id*='toDate']")
                to_field.clear()
                to_field.send_keys(to_str)

                # Välj ämnesområde om angivet
                if amnesomrade:
                    try:
                        select_elem = self.driver.find_element(By.CSS_SELECTOR,
                            "select[name*='amne'], select[id*='amnesomrade']")
                        select = Select(select_elem)
                        select.select_by_visible_text(amnesomrade)
                    except NoSuchElementException:
                        logger.warning(f"Kunde inte välja ämnesområde: {amnesomrade}")

                # Sök
                search_btn = self.driver.find_element(By.CSS_SELECTOR,
                    "button[type='submit'], input[type='submit']")
                search_btn.click()

                time.sleep(2)
                results = self._parse_search_results()

            except TimeoutException:
                logger.warning("Kunde inte hitta datumfält")
            except NoSuchElementException as e:
                logger.warning(f"Element saknas: {e}")

        except Exception as e:
            logger.error(f"Fel vid datumsökning: {e}")

        return results

    def _parse_search_results(self) -> List[Kungorelse]:
        """Parsar sökresultat från POIT"""
        results = []

        try:
            # Vänta på resultatlista
            wait = WebDriverWait(self.driver, 10)

            # POIT visar resultat i en tabell eller lista
            # Försök hitta resultatrader
            result_rows = self.driver.find_elements(By.CSS_SELECTOR,
                "table.results tr, .result-item, .kungorelse-item, ul.results li")

            for row in result_rows:
                try:
                    kungorelse = self._parse_result_row(row)
                    if kungorelse:
                        results.append(kungorelse)
                except Exception as e:
                    logger.debug(f"Kunde inte parsa rad: {e}")

            logger.info(f"Hittade {len(results)} kungörelser")

        except TimeoutException:
            logger.info("Inga resultat hittades")
        except Exception as e:
            logger.error(f"Fel vid parsing av resultat: {e}")

        return results

    def _parse_result_row(self, row) -> Optional[Kungorelse]:
        """Parsar en enskild resultatrad"""
        try:
            # Försök extrahera data från raden
            # OBS: Exakta selektorer behöver anpassas efter POIT:s faktiska HTML

            text = row.text.strip()
            if not text:
                return None

            # Försök hitta länk till kungörelsen
            link = None
            try:
                link_elem = row.find_element(By.TAG_NAME, "a")
                link = link_elem.get_attribute("href")
            except NoSuchElementException:
                pass

            # Extrahera kungörelse-ID från länk eller text
            kungorelse_id = ""
            if link and "kungorelse" in link.lower():
                # Extrahera ID från URL
                parts = link.split("/")
                kungorelse_id = parts[-1] if parts else ""

            # Skapa kungörelse-objekt
            return Kungorelse(
                kungorelse_id=kungorelse_id or f"temp_{hash(text)}",
                rubrik=text[:200],  # Första 200 tecken som rubrik
                amnesomrade="Okänt",
                publiceringsdatum=datetime.now().strftime("%Y-%m-%d"),
                url=link
            )

        except Exception as e:
            logger.debug(f"Fel vid parsing av rad: {e}")
            return None

    def get_kungorelse_details(self, kungorelse_url: str) -> Optional[Dict[str, Any]]:
        """Hämtar detaljerad information om en kungörelse"""
        self.start()

        try:
            self.driver.get(kungorelse_url)

            if not self._wait_for_page_load():
                return None

            # Extrahera all text från sidan
            content = self.driver.find_element(By.TAG_NAME, "body").text

            # Försök extrahera strukturerad data
            details = {
                "url": kungorelse_url,
                "content": content,
                "fetched_at": datetime.now().isoformat()
            }

            # Sök efter organisationsnummer i texten
            import re
            orgnr_match = re.search(r'\b(\d{6}-?\d{4})\b', content)
            if orgnr_match:
                details["organisationsnummer"] = orgnr_match.group(1).replace("-", "")

            return details

        except Exception as e:
            logger.error(f"Fel vid hämtning av detaljer: {e}")
            return None


# Enkel test
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)

    with POITScraper(headless=False) as scraper:
        # Testa sökning
        from_date = datetime.now() - timedelta(days=7)
        to_date = datetime.now()

        results = scraper.search_by_date_range(from_date, to_date)

        for r in results[:5]:
            print(f"- {r.rubrik[:50]}...")

File: src/scrapers/poit-scraper.js
/**
 * POIT Scraper - Node.js version med puppeteer-extra stealth
 * Hämtar kungörelser från Post- och Inrikes Tidningar
 *
 * Använder centraliserade moduler:
 * - browser-factory: Browser-skapande med stealth, adblocker, CAPTCHA-hantering
 * - popup-blocker: Cookie consent, popup-hantering
 */

const {
    createBrowser,
    createPage,
    configurePage,
    dismissAllPopups,
    sleep
} = require('../utils/browser-factory');

/**
 * Parsar sökresultat från POIT-sidan
 */
async function parseSearchResults(page) {
    return await page.evaluate(() => {
        const results = [];
        const rows = document.querySelectorAll('table tbody tr, .result-row');

        rows.forEach(row => {
            const cells = row.querySelectorAll('td');
            if (cells.length >= 5) {
                const linkElement = row.querySelector('a');
                results.push({
                    kungorelse_id: cells[0]?.textContent?.trim() || '',
                    uppgiftslamnare: cells[1]?.textContent?.trim() || '',
                    typ: cells[2]?.textContent?.trim() || '',
                    namn: cells[3]?.textContent?.trim() || '',
                    publicerad: cells[4]?.textContent?.trim() || '',
                    url: linkElement?.href || null
                });
            }
        });

        // Hämta antal träffar
        const antalText = document.body.innerText.match(/Antal träffar:\s*(\d+)/);
        const antalTraffar = antalText ? parseInt(antalText[1]) : results.length;

        return {
            antal_traffar: antalTraffar,
            kungorelser: results
        };
    });
}

/**
 * Söker kungörelser för ett organisationsnummer
 */
async function searchByOrgnr(orgnr, options = {}) {
    const { headless = true, timeout = 60000 } = options;

    // Använd centraliserad browser-factory
    const browser = await createBrowser({ headless });
    const page = await createPage(browser);

    try {
        // Navigera till POIT
        await page.goto('https://poit.bolagsverket.se/poit-app/', {
            waitUntil: 'domcontentloaded',
            timeout: timeout
        });
        await sleep(3000);

        // Konfigurera sidan och stäng popups/cookies
        await configurePage(page);
        await dismissAllPopups(page);

        // Gå till söksidan
        await page.evaluate(() => {
            const link = document.querySelector('a[href="/poit-app/sok"]');
            if (link) link.click();
        });
        await sleep(3000);

        // Stäng eventuella nya popups
        await dismissAllPopups(page);

        // Fyll i organisationsnummer
        await page.waitForSelector('#personOrgnummer', { timeout: 10000 });
        await page.type('#personOrgnummer', orgnr.replace(/-/g, ''), { delay: 30 });
        await sleep(2000);

        // Klicka på sök
        const clicked = await page.evaluate(() => {
            const buttons = Array.from(document.querySelectorAll('button'));
            const searchBtn = buttons.find(b =>
                b.textContent.includes('Sök kungörelse') && !b.disabled
            );
            if (searchBtn) {
                searchBtn.click();
                return true;
            }
            return false;
        });

        if (!clicked) {
            throw new Error('Kunde inte hitta aktiv sökknapp');
        }

        await sleep(5000);

        // Parsa resultat
        const results = await parseSearchResults(page);

        return {
            success: true,
            orgnr: orgnr,
            ...results
        };

    } catch (error) {
        return {
            success: false,
            orgnr: orgnr,
            error: error.message,
            antal_traffar: 0,
            kungorelser: []
        };
    } finally {
        await browser.close();
    }
}

/**
 * Söker kungörelser för flera organisationsnummer
 */
async function searchMultiple(orgnrList, options = {}) {
    const results = [];

    for (const orgnr of orgnrList) {
        console.error(`Söker: ${orgnr}`);
        const result = await searchByOrgnr(orgnr, options);
        results.push(result);

        // Vänta lite mellan sökningar
        await sleep(2000);
    }

    return results;
}

/**
 * Hämtar detaljerad information om en specifik kungörelse
 * @param {string} kungorelseId - Kungörelse-ID (t.ex. "K967902/25" eller "K967902-25")
 * @param {object} options - Alternativ (headless, timeout)
 * @returns {object} Detaljerad kungörelse-information
 */
async function getKungorelseDetails(kungorelseId, options = {}) {
    const { headless = true, timeout = 60000 } = options;

    // Normalisera ID: K967902/25 -> K967902-25
    const normalizedId = kungorelseId.replace('/', '-');
    const url = `https://poit.bolagsverket.se/poit-app/kungorelse/${normalizedId}`;

    // Använd centraliserad browser-factory
    const browser = await createBrowser({ headless });
    const page = await createPage(browser);

    try {
        console.error(`Hämtar kungörelse: ${url}`);

        await page.goto(url, {
            waitUntil: 'networkidle2',
            timeout: timeout
        });
        await sleep(3000);

        // Konfigurera sidan och stäng popups/cookies
        await configurePage(page);
        await dismissAllPopups(page);
        await sleep(1000);

        // Extrahera all information från sidan
        const details = await page.evaluate(() => {
            const result = {
                kungorelseText: '',
                typ: '',
                uppgiftslamnare: '',
                foretag: '',
                orgnummer: '',
                datum: '',
                forvaltare: null,
                adress: null,
                telefon: null,
                epost: null,
                domstol: null,
                diarienummer: null,
                andringar: [],
                rawText: ''
            };

            // Hämta hela texten på sidan
            const bodyText = document.body.innerText;
            result.rawText = bodyText;

            // Försök hitta huvudinnehållet
            const mainContent = document.querySelector('main, .content, .kungorelse-content, article');
            if (mainContent) {
                result.kungorelseText = mainContent.innerText.trim();
            } else {
                result.kungorelseText = bodyText;
            }

            // Extrahera organisationsnummer (format: 556890-8288)
            const orgnrMatch = bodyText.match(/\b(\d{6}-\d{4})\b/);
            if (orgnrMatch) {
                result.orgnummer = orgnrMatch[1];
            }

            // Extrahera företagsnamn (vanligtvis före orgnumret)
            if (result.orgnummer) {
                const beforeOrgnr = bodyText.split(result.orgnummer)[0];
                const lines = beforeOrgnr.split('\n').filter(l => l.trim());
                // Ta sista icke-tomma raden före orgnumret
                for (let i = lines.length - 1; i >= 0; i--) {
                    const line = lines[i].trim();
                    if (line.length > 3 && line.length < 200 && !line.match(/^\d/) && !line.includes('Kungörelse')) {
                        result.foretag = line;
                        break;
                    }
                }
            }

            // Extrahera förvaltare (konkursfall)
            const forvaltareMatch = bodyText.match(/[Ff]örvaltare\s*(?:är|:)?\s*(?:advokat\s*)?([^,\n]+?)(?:,|\.|telefon|tel|$)/i);
            if (forvaltareMatch) {
                result.forvaltare = forvaltareMatch[1].trim();
            }

            // Extrahera telefonnummer
            const telefonMatch = bodyText.match(/(?:telefon|tel\.?|tfn\.?)[\s:]*([0-9\s\-+()]{8,20})/i);
            if (telefonMatch) {
                result.telefon = telefonMatch[1].trim();
            }

            // Extrahera e-post
            const epostMatch = bodyText.match(/([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})/);
            if (epostMatch) {
                result.epost = epostMatch[1];
            }

            // Extrahera domstol/tingsrätt
            const domstolMatch = bodyText.match(/(\w+\s+tingsrätt)/i);
            if (domstolMatch) {
                result.domstol = domstolMatch[1];
            }

            // Extrahera datum
            const datumMatch = bodyText.match(/(\d{4}-\d{2}-\d{2})/);
            if (datumMatch) {
                result.datum = datumMatch[1];
            }

            // Bestäm typ av kungörelse
            const lowerText = bodyText.toLowerCase();
            if (lowerText.includes('konkursbeslut') || lowerText.includes('försatt i konkurs')) {
                result.typ = 'Konkursbeslut';
            } else if (lowerText.includes('likvidation')) {
                result.typ = 'Likvidation';
            } else if (lowerText.includes('fusion')) {
                result.typ = 'Fusion';
            } else if (lowerText.includes('kallelse') && lowerText.includes('borgenär')) {
                result.typ = 'Kallelse på okända borgenärer';
            } else if (lowerText.includes('aktiebolagsregistret')) {
                result.typ = 'Aktiebolagsregistret';
            } else if (lowerText.includes('nyemission') || lowerText.includes('aktiekapital')) {
                result.typ = 'Nyemission';
            } else if (lowerText.includes('styrelse')) {
                result.typ = 'Styrelseändring';
            } else if (lowerText.includes('bolagsordning')) {
                result.typ = 'Bolagsordningsändring';
            } else {
                result.typ = 'Kungörelse';
            }

            // Extrahera ändringar (för Aktiebolagsregistret)
            const andringarSection = bodyText.match(/Ändringar som har registrerats[:\s]*([^]*?)(?=Uppgiftslämnare|$)/i);
            if (andringarSection) {
                const andringarText = andringarSection[1];
                const andringarLines = andringarText.split('\n').map(l => l.trim()).filter(l => l.length > 0);
                result.andringar = andringarLines;
            }

            // Extrahera uppgiftslämnare
            const uppgiftslamnareMatch = bodyText.match(/Uppgiftslämnare[:\s]*([^\n]+)/i);
            if (uppgiftslamnareMatch) {
                result.uppgiftslamnare = uppgiftslamnareMatch[1].trim();
            }

            return result;
        });

        return {
            success: true,
            kungorelse_id: normalizedId,
            url: url,
            ...details,
            fetched_at: new Date().toISOString()
        };

    } catch (error) {
        return {
            success: false,
            kungorelse_id: normalizedId,
            url: url,
            error: error.message
        };
    } finally {
        await browser.close();
    }
}

/**
 * Hämtar detaljer för flera kungörelser
 * @param {string[]} kungorelseIds - Lista med kungörelse-IDs
 * @param {object} options - Alternativ
 * @returns {object[]} Lista med detaljerade kungörelser
 */
async function getMultipleKungorelseDetails(kungorelseIds, options = {}) {
    const results = [];

    for (const id of kungorelseIds) {
        console.error(`Hämtar detaljer för: ${id}`);
        const result = await getKungorelseDetails(id, options);
        results.push(result);

        // Vänta lite mellan förfrågningar
        await sleep(2000);
    }

    return results;
}

// CLI-läge - utökat med --details flagga
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length === 0) {
        console.error('Användning:');
        console.error('  Sökning:  node poit-scraper.js <orgnr> [--visible]');
        console.error('  Detaljer: node poit-scraper.js --details <kungörelse-id> [--visible]');
        console.error('');
        console.error('Exempel:');
        console.error('  node poit-scraper.js 5593220048');
        console.error('  node poit-scraper.js --details K967902-25');
        process.exit(1);
    }

    const headless = !args.includes('--visible');
    const isDetails = args.includes('--details');

    if (isDetails) {
        const detailsIndex = args.indexOf('--details');
        const kungorelseId = args[detailsIndex + 1];

        if (!kungorelseId || kungorelseId.startsWith('--')) {
            console.error('Fel: Ange kungörelse-ID efter --details');
            process.exit(1);
        }

        getKungorelseDetails(kungorelseId, { headless })
            .then(result => {
                console.log(JSON.stringify(result, null, 2));
            })
            .catch(err => {
                console.error('Fel:', err.message);
                process.exit(1);
            });
    } else {
        const orgnr = args[0];

        searchByOrgnr(orgnr, { headless })
            .then(result => {
                console.log(JSON.stringify(result, null, 2));
            })
            .catch(err => {
                console.error('Fel:', err.message);
                process.exit(1);
            });
    }
}

module.exports = {
    searchByOrgnr,
    searchMultiple,
    getKungorelseDetails,
    getMultipleKungorelseDetails
};
